ports/unix/mphciport.c

```diff
@@ -0,0 +1,198 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/runtime.h"
+#include "py/mperrno.h"
+#include "py/mphal.h"
+
+#if MICROPY_PY_BLUETOOTH && (MICROPY_BLUETOOTH_NIMBLE || (MICROPY_BLUETOOTH_BTSTACK && MICROPY_BLUETOOTH_BTSTACK_H4))
+
+#if !MICROPY_PY_THREAD
+#error Unix HCI UART requires MICROPY_PY_THREAD
+#endif
+
+#include "extmod/modbluetooth.h"
+#include "extmod/mphciport.h"
+
+#include <pthread.h>
+#include <unistd.h>
+
+#include <termios.h>
+#include <fcntl.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define DEBUG_printf(...) // printf(__VA_ARGS__)
+#define DEBUG_HCI_DUMP (0)
+
+uint8_t mp_bluetooth_hci_cmd_buf[4 + 256];
+
+// Must be provided by the stack bindings (e.g. mpnimbleport.c or mpbtstackport.c).
+extern bool mp_bluetooth_hci_poll(void);
+
+STATIC const useconds_t UART_POLL_INTERVAL_US = 1000;
+
+STATIC int uart_fd = -1;
+STATIC pthread_t hci_poll_thread_id;
+
+STATIC void *hci_poll_thread(void *arg) {
+    (void)arg;
+
+    // This will return false when the stack is shutdown.
+    while (mp_bluetooth_hci_poll()) {
+        usleep(UART_POLL_INTERVAL_US);
+    }
+
+    return NULL;
+}
+
+STATIC int configure_uart(void) {
+    struct termios toptions;
+
+    // Get existing config.
+    if (tcgetattr(uart_fd, &toptions) < 0) {
+        DEBUG_printf("Couldn't get term attributes");
+        return -1;
+    }
+
+    // Raw mode (disable all processing).
+    cfmakeraw(&toptions);
+
+    // 8N1, no parity.
+    toptions.c_cflag &= ~CSTOPB;
+    toptions.c_cflag |= CS8;
+    toptions.c_cflag &= ~PARENB;
+
+    // Enable receiver, ignore modem control lines
+    toptions.c_cflag |= CREAD | CLOCAL;
+
+    // Blocking, single-byte reads.
+    toptions.c_cc[VMIN] = 1;
+    toptions.c_cc[VTIME] = 0;
+
+    // Enable HW RTS/CTS flow control.
+    toptions.c_iflag &= ~(IXON | IXOFF | IXANY);
+    toptions.c_cflag |= CRTSCTS;
+
+    // 1Mbit (TODO: make this configurable).
+    speed_t brate = B1000000;
+    cfsetospeed(&toptions, brate);
+    cfsetispeed(&toptions, brate);
+
+    // Apply immediately.
+    if (tcsetattr(uart_fd, TCSANOW, &toptions) < 0) {
+        DEBUG_printf("Couldn't set term attributes");
+        return -1;
+    }
+
+    return 0;
+}
+
+// HCI UART bindings.
+int mp_bluetooth_hci_uart_init(uint32_t port, uint32_t baudrate) {
+    (void)port;
+    (void)baudrate;
+
+    DEBUG_printf("mp_bluetooth_hci_uart_init (unix)\n");
+
+    char uart_device_name[256] = "/dev/ttyUSB0";
+
+    char *path = getenv("MICROPYBTUART");
+    if (path != NULL) {
+        strcpy(uart_device_name, path);
+    }
+    DEBUG_printf("Using HCI UART: %s\n", uart_device_name);
+
+    int flags = O_RDWR | O_NOCTTY | O_NONBLOCK;
+    uart_fd = open(uart_device_name, flags);
+    if (uart_fd == -1) {
+        DEBUG_printf("Unable to open port %s", uart_device_name);
```

Comment:
This would be better as an exception that's then user visible - currently if the port is missing or cannot be opened it's a silent fail here and the rest of the startup just looks broken

Reply:
Done -- and added the plumbing to make this work for btstack and nimble. On btstack it can detect the init failure, on nimble it now leads to an init timeout.